Sign validation request when using aws-sigv4 (#130)

robinroestenburg opened the pull request:
Signed-off-by: Robin Roestenburg <robin.roestenburg@4me.com>

### Description
Fixes #130 by having the `opensearch-aws-sigv4` gem pass the signing headers to the request to verify / validate the OpenSearch distribution. 

When you do not use the AWS v4 signing, the behaviour of the `opensearch-ruby` gem is unchanged. The `verify_open_search` method is called without passing any headers (as before).

### Issues Resolved
Closes #130. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock commented on opensearch-aws-sigv4/lib/opensearch-aws-sigv4.rb:
```diff
@@ -50,7 +50,12 @@ def initialize(transport_args = {}, sigv4_signer, &block)
+            http_method: 'GET',
+            url: signature_url('/', {}))
+          verify_open_search(verify_signature.headers)
+        end
```
Refactor this. 

1. Implement `verify_open_search` as a private method in this class in which you do the `verify_signature = `..., then call `super`. 
1. Same for below, instead of calling `@transport.perform_request`, call `super` after generating the signature.

robinroestenburg replied:
@dblock thanks for your feedback, I've updated the PR with your suggested refactoring.

dblock reviewed (changes requested):
Thanks! Let's fix the method override? Anyone calling `verify_open_search` should be getting a signed request, not just when doing `perform_request`.

nhtruong commented on opensearch-aws-sigv4/lib/opensearch-aws-sigv4.rb:
```diff
@@ -45,24 +45,30 @@ def initialize(transport_args = {}, sigv4_signer, &block)
+        verify_signature = sigv4_signer.sign_request(
+          http_method: 'GET',
+          url: signature_url('/', {}))
+        super(verify_signature.headers)
```
Wouldn't this error out because `OpenSearch::Client#verify_open_search`, the super method of this one, doesn't accept any params?

I think the only change needed is overriding `open_search_validation_request` method:
```ruby
def open_search_validation_request
  signature = sigv4_signer.sign_request(http_method: 'GET', url: signature_url('/', {}))
  @transport.perform_request('GET', '/', {}, nil, signature.headers)
end
```

nhtruong commented:
I think the only change needed is overriding `open_search_validation_request` method:
```ruby
def open_search_validation_request
  method = 'GET'
  path = '/'
  signature = sigv4_signer.sign_request(http_method: method, url: signature_url(path, {}))
  @transport.perform_request(method, path, {}, nil, signature.headers)
end
```
Though maybe I'm missing sometlhing.

nhtruong commented:
This also need a rebase. Each gem now has its own changelog inside its respective folder.

robinroestenburg commented:
> I think the only change needed is overriding open_search_validation_request method since the verification logic is virtually identical to the base client

@nhtruong yes, I think you are right. I have adjusted the code (and rebased onto the latest version).
